ragna/deploy/_config.py

```diff
@@ -76,14 +75,48 @@ def settings_customise_sources(
         #  4. Default
         return env_settings, init_settings
 
+    local_root: Annotated[Path, AfterValidator(make_directory)] = Field(
+        default_factory=ragna.local_root
+    )
+
+    authentication: ImportString[type[Authentication]] = (
+        "ragna.deploy.RagnaDemoAuthentication"  # type: ignore[assignment]
+    )
+
+    document: ImportString[type[Document]] = "ragna.core.LocalDocument"  # type: ignore[assignment]
+    source_storages: list[ImportString[type[SourceStorage]]] = [
+        "ragna.source_storages.RagnaDemoSourceStorage"  # type: ignore[list-item]
+    ]
+    assistants: list[ImportString[type[Assistant]]] = [
+        "ragna.assistants.RagnaDemoAssistant"  # type: ignore[list-item]
+    ]
+
+    hostname: str = "127.0.0.1"
+    port: int = 31476
+    root_path: str = ""
+    origins: list[str] = AfterConfigValidateDefault.make(
+        lambda config: [f"http://{config.hostname}:{config.port}"]
+    )
+
+    database_url: str = AfterConfigValidateDefault.make(
+        lambda config: f"sqlite:///{config.local_root}/ragna.db",
+    )
+
+    @model_validator(mode="after")
+    def _validate_model(self) -> Config:
+        self._resolve_default_sentinels(self)
+        return self
+
     def _resolve_default_sentinels(self, config: Config) -> None:
         for name, info in self.model_fields.items():
             value = getattr(self, name)
-            if isinstance(value, ConfigBase):
-                value._resolve_default_sentinels(config)
-            elif isinstance(value, AfterConfigValidateDefault):
+            if isinstance(value, AfterConfigValidateDefault):
                 setattr(self, name, value.make_default(config))
 
+    @property
+    def _url(self) -> str:
+        return f"http://{self.hostname}:{self.port}{self.root_path}"
+
```

Comment:
Should I be able to specify a custom root path in the config file? When I try something like
> root_path = "/test"

I get 404s everywhere. Am I misunderstanding what this does or should this root path be passed along to the router?

Reply:
TL;DR: You are seeing 404, because you set the `root_path` option, but you are not behind a proxy that enforces this.

The `root_path` config option is needed when you run Ragna behind a proxy, e.g. [`jupyter-server-proxy`](https://pypi.org/project/jupyter-server-proxy/). What this option does is to automatically handle the routing for you inside our code.

Imagine Ragna is deployed behind a proxy and users can access it at `my-company.com/ragna`. In this configuration, the `root_path` would be `/ragna`. When we now define an endpoint as `@app.post("/foo")`, FastAPI automatically routes any request to `/ragna/foo` to the endpoint. Meaning, the `root_path` is an abstraction so our application is agnostic to any potential proxy in front of it.

 See the [config reference](
https://ragna.chat/en/stable/references/config/#root_path). If you think the documentation is not sufficient, feel free to send a PR to improve it.